support individual registry keys for file settings

adipose opened the pull request:
This is a starting point for storing file settings in individual reg keys.

1. Used built-in windows crytpo library with sha256.  I only use 10 chars of the base64 encoding, which is very likely to be unique.  Good enough for our purposes. std::hash is not guaranteed to have same results across processes, so I didn't use it.
2. Small amount of backwards compatibility--it can load the existing file history list, but doesn't retain title, cue, subs.  I didn't think that was super important to keep around.
3. It keeps file settings, even if they drop out of the history.  I thought this was a nice feature.  File history needs a limit, but settings can stay a long time.  But maybe it should clean up sometimes.

Other features like A-B could be added here.

adipose committed:
support individual registry keys for file settings

clsid2 commented:
In getFilenameHash(), make the filename lowercase (except if URL which are partly case sensitive).

For backwards compatibility, we could simply check if the new registry key exists, and if not, run the old code.

My idea was to avoid using "hash001"/"hash002"/etc values. So just a subkey for each history entry.

Key:
`HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\History\[FILEHASH]`
Values:
timestamp of last update
filename (or maybe something like "location" is better name?)
etc.

Then when reading the data, we enumerate all subkeys in History key. Each subkey is a history entry. We insert them into a sorted list based on the timestamp.
When reading is done, simply check if list size exceeds our maximum, and remove the oldest entries (and the corresponding subkeys). We could even prune entries that have a certain age.

For reading old list, the timestamp could be set to current time for first entry, and then substract 1 for each next one, to keep correct sorting.

ale5000-git commented:
Also it would be nice to normalize URLs (to avoid duplicate entries).

These are the same:
```
www.domain.com
www.domain.com/
WWW.DOMAIN.COM
```

While these are all different:
```
www.domain.com/folder
www.domain.com/folder/
www.domain.com/FOLDER/
```

adipose commented:
> 
> 
> In getFilenameHash(), make the filename lowercase (except if URL which are partly case sensitive).
> 
> For backwards compatibility, we could simply check if the new registry key exists, and if not, run the old code.
> 
> My idea was to avoid using "hash001"/"hash002"/etc values. So just a subkey for each history entry.
> 
> Key: `HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\History\[FILEHASH]` Values: timestamp of last update filename (or maybe something like "location" is better name?) etc.
> 
> Then when reading the data, we enumerate all subkeys in History key. Each subkey is a history entry. We insert them into a sorted list based on the timestamp. When reading is done, simply check if list size exceeds our maximum, and remove the oldest entries (and the corresponding subkeys). We could even prune entries that have a certain age.
> 
> For reading old list, the timestamp could be set to current time for first entry, and then substract 1 for each next one, to keep correct sorting.

Yes, I did consider doing it that way per your suggestion.  I ended up keeping the list of 20 items because it was simpler to design, and allows a longer list of save file settings.  I found it convenient to be able to see the list of hashes in an order--without that list, you have an unordered list of hashes which you'd have to dig through manually (I realize this is not something the average user would do, but it's convenient when testing things).  One option might be to have a single SZ with the list of hashes delimited.

If we have just the keys, we load all of them, sort them, and then throw away those > 20th entry (purge the key).  Or, we just leave them there, but don't show them in recent files--but the list of keys to sort would grow and make loading the player slower.

clsid2 commented:
If the list becomes to big, then yes, simply delete the oldest keys.

I really prefer the design without the index list. Because then we can update the contents of an individual history entry without having to touch any other keys.

adipose commented:
Oh, right.

adipose committed:
migration for old file history. write media history on file open

adipose committed:
Merge remote-tracking branch 'upstream/develop' into patch228

adipose commented:
@clsid2 

I think this is working.  I tested with two instances and the history is successfully shared across them.

Migration of old file history is working.

Did nothing with the "Dub" code.  It's similar to the old file history code.

Nothing implemented to clean up old entries.  I suggest an advanced option to limit the number of retained entries (default 100).

adipose commented:
> 
> 
> Also it would be nice to normalize URLs (to avoid duplicate entries).
> 
> These are the same:
> 
> ```
> www.domain.com
> www.domain.com/
> WWW.DOMAIN.COM
> ```
> 
> While these are all different:
> 
> ```
> www.domain.com/folder
> www.domain.com/folder/
> www.domain.com/FOLDER/
> ```

This seems like a valid improvement to the hash.  However, given that the list length will be limited, how likely is it the same file is loaded with multiple, redundant paths?  In the event that it is, it will be a nearly harmless result that subs or place will have to be rechosen.

Feel free to submit a patch to improve it, though.

clsid2 commented:
I will look at this in more detail later.

There already is an advanced option for the number of history entries. We could do the cleanup when reading the list. If too many entries -> remove oldest ones and their corresponding keys.
At this point it is not important to maintain max size during an active session. That can always be added later.

The whole list only needs to be read once.
When opening a file, read the key for that hash again if it exists. That way the file info is up-to-date.

We also don't need to write the whole list. Just the individual entries when they get modified.

Ideally the playback position data should be moved to this new scheme as well.

adipose commented:
It should only write the list once. Otherwise it gets updated in the add and remove calls.

Agreed about the playback position. I will add that.

adipose committed:
remove unnecessary list reads

adipose committed:
Merge remote-tracking branch 'upstream/develop' into patch228

vlakoff commented:
Maybe I would suggest slightly raising the cropped base64 hash, from 10 to 12 characters.

10 characters has 60-bit entropy, so it's a bit below a "psychological" value of 64-bit entropy (for perspective, MD5 hashes are 128-bit). Raising to 12 characters would make us reach 72-bit entropy, and go above the 64-bit entropy.

(it's just a suggestion, such discussions could get endless… both values are fine, let clsid make the pick)

adipose committed:
integrate fileposition and DVDposition to mediahistory

adipose commented:
1. Updated hash length to 12 characters.  It hardly matters what the length is.  It could easily be longer with no real impact on ease of use.  I figured it doesn't need to be very high entropy for 40 files of history.
2. Doesn't remove old entries yet (but only loads latest 40/history limit).
3. Fixed a bug in MainFrm.cpp where it is supposed to persist every 30 seconds, but I think the code wouldn't have ever fired (tries to find a 30 second difference, but it updates the current value every event, so 30 second diff never occurs).
4. Integrated file and dvd positions.
5. Eliminate current_rfe var in favor of looking up the current one.  Otherwise, it just has to be updated every time the existing list is updated.
6. Eliminate mediapositionlist project references (code remains for now).

clsid2 commented:
In case of AddSubToCurrent() and UpdateCurrentFilePosition() maybe add an optional boolean parameter "save" that if true calls WriteCurrentEntry() when relevant.

adipose commented:
> In case of AddSubToCurrent() and UpdateCurrentFilePosition() maybe add an optional boolean parameter "save" that if true calls WriteCurrentEntry() when relevant.

I considered it.  For UpdateCurrentFilePosition(), we can use with diff > 30 seconds.  For AddSubToCurrent(), it will get saved when the program closes anyway.

clsid2 commented:
Saving the position in OnTimer() could be increased to once per 60sec. That position save is only relevant in case of a crash, which should be a rare occurrence.
But bug still exists? The value is still updated each time unless I read code wrong? Shouldn't the value update be under `if (bSave)`?

adipose commented:
> Saving the position in OnTimer() could be increased to once per 60sec. That position save is only relevant in case of a crash, which should be a rare occurrence.
> 
> But bug still exists? The value is still updated each time unless I read code wrong? Shouldn't the value update be under `if (bSave)`?

I need to fix it again. After latest changes I reintroduced the bug.

It needs to always set the value so it will be correct in the RFE if written later. But the 30 second interval needs to be measured a different way.

adipose committed:
use 12 character hash

adipose committed:
fix force persist after 30 seconds

adipose commented:
Should be fixed, now.  I still need to add a cleanup section for old media hashes.

clsid2 commented on src/mpc-hc/MainFrm.cpp:
```diff
@@ -1978,11 +1978,7 @@ void CMainFrame::OnTimer(UINT_PTR nIDEvent)
                             if (m_bRememberFilePos && !m_fEndOfStream) {
                                 auto* pMRU = &AfxGetAppSettings().MRU;
```
Since persistedFilePosition is only relevant for current playing file, I would be more efficient to store it as a general value in the appsettings.

adipose committed:
single var for persistedFilePosition

adipose committed:
fix bug with files opened through OpenMediaPrivate

adipose commented:
@clsid2, I think all features are working and complete, including cleaning up old reg keys.

It might be worth considering a separate constant for max fileposition history.  As it was previously implemented, they were stored separately and you could have as many as double (due to DVD and file positions being stored separately).  But it could be convenient to have it store 100+ file positions, while only showing the latest 40 in the "recent files."  That way if you open an old video it will remember the position, even if it wasn't recently opened.

This code might not work for .ini files, I just remembered.

clsid2 commented:
Recent files list could be maximized to 40. I doubt many will seek down in a larger list.
The existing RecentFilesNumber option can be used for the history count.

INI support needs to be added before merging.

P.S. I might do a 1.9.19 release coming week. Are there are useful Transifex additions?

adipose committed:
support .ini files with subkeys

adipose commented:
I need to review transifex.  I added .ini support just now.

adipose committed:
only load direct subkeys from ini file

clsid2 commented:
Will probably merge this after upcoming release. That way we can have some proper testing in a dev build first just in case some regressions might turn up.

adipose commented:
I agree, it needs testing.  I've tested quite a few scenarios, but some things that it does:

1. Stores history
2. Stores location
3. Stores dvd location
4. Hashes dvds by the dvd id instead of filename
5. Migrates old settings and clears them
6. Loads dynamically to support multiple instances

All those things could be tested to make sure they behave as expected.

clsid2 commented:
Can the hash value contain a backslash character? If so, we need to replace that to avoid issues with the reg/ini subkey name.

adipose commented:
I believe base64 only has a forward slash. Also a plus sign.

https://en.wikipedia.org/wiki/Base64

vlakoff commented:
I noticed PowerShell has issues with "/" in registry key name ([this Google search](https://www.google.com/search?q=registry+key+slash) points to many topics, like [this one](https://stackoverflow.com/questions/18218835/how-to-create-a-registry-entry-with-a-forward-slash-in-the-name)).

Thus, other tools/API _might_ have similar issues (though I doubt it and I guess only PowerShell has such issue).
Maybe do an attempt with a hash containing "/" just to make sure it works?

adipose commented:
![image](https://user-images.githubusercontent.com/3324395/152664822-563f3dc0-71b2-4498-b043-0f6c5f3de9df.png)

I have seen it already several times. No problems so far.  Nobody should need to hardcode the key names anyway in a PS script, but we could replace it with something else.
